src/main/java/org/opensearch/knn/index/codec/nativeindex/NativeIndexWriterScratchIter.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.codec.nativeindex;
+
+import java.io.IOException;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.util.Map;
+
+import org.apache.lucene.index.BinaryDocValues;
+import org.opensearch.knn.index.codec.util.KNNCodecUtil;
+import org.opensearch.knn.index.util.KNNEngine;
+import org.opensearch.knn.jni.JNIService;
+
+import lombok.extern.log4j.Log4j2;
+
+/**
+ * Class to build the KNN index from scratch iteratively and write it to disk
+ */
+@Log4j2
+public class NativeIndexWriterScratchIter extends NativeIndexWriterScratch {
+
+    @Override
+    protected void createIndex(NativeIndexInfo indexInfo, BinaryDocValues values) throws IOException {
+        long indexAddress = initIndexFromScratch(
+            indexInfo.getNumDocs(),
+            indexInfo.getVectorInfo().getDimension(),
+            indexInfo.getKnnEngine(),
+            indexInfo.getParameters()
+        );
+        while (true) {
+            KNNCodecUtil.VectorBatch batch = KNNCodecUtil.getVectorBatch(
+                values,
+                getVectorTransfer(indexInfo.getVectorInfo().getVectorDataType()),
+                true
+            );
+            insertToIndex(batch, indexInfo.getKnnEngine(), indexAddress, indexInfo.getParameters());
+            if (batch.finished) {
+                break;
+            }
+        }
```

Comment:
instead of having a while loop @MrFlap do you think if we can find out how trips to JNI will be required and then we use that as counter/terminating condition for breaking out of the loop? while(true) is giving me scares of infinite loop.

Reply:
Right. I don't think there is a great way to do this since different batches could have different numbers of documents. The most maintainable way may be to accumulate the number of docs until we've reached totalLiveDocsCount?

Reply:
Make sense. Not blocking this PR because of this change,